iree/compiler/Conversion/HLOToLinalg/HLOToLinalgOnTensors.cpp

```diff
@@ -333,18 +335,91 @@ LogicalResult DepthwiseConvOpConversion::matchAndRewrite(
+  using OpConversionPattern<linalg::PadTensorOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      linalg::PadTensorOp padTensorOp, ArrayRef<Value> args,
+      ConversionPatternRewriter &rewriter) const override {
+    linalg::PadTensorOpAdaptor padOpAdaptor(args,
+                                            padTensorOp->getAttrDictionary());
+    // Check that the region is just a yield operation which is returning a
+    // scalar that is not one of the arguments of the linalg operation.
+    Region &region = padTensorOp.region();
+    Block &block = region.front();
+    if (!llvm::hasSingleElement(block)) return failure();
+    auto yieldOp = cast<linalg::YieldOp>(block.getTerminator());
+    if (!llvm::hasSingleElement(yieldOp.values())) return failure();
+    Value yieldVal = yieldOp.values().front();
+    if (llvm::any_of(block.getArguments(),
+                     [&](Value v) { return v == yieldVal; })) {
+      return failure();
+    }
+
+    OpBuilder::InsertionGuard g(rewriter);
+    Location loc = padTensorOp.getLoc();
+    auto lowPadVals = padTensorOp.getLowPadAsValues(rewriter, loc);
+    auto highPadVals = padTensorOp.getHighPadAsValues(rewriter, loc);
+    Value source = padOpAdaptor.source();
+    RankedTensorType sourceType = padTensorOp.getSourceType();
+    int64_t rank = sourceType.getRank();
+
+    // TODO(ravishankarm): Use shape inference interface to get this.
+    auto sourceShape = llvm::to_vector<4>(
+        llvm::map_range(llvm::seq<int64_t>(0, rank), [&](int64_t dim) -> Value {
+          return rewriter.create<DimOp>(loc, source, dim);
+        }));
+    auto outputShape = llvm::to_vector<4>(
+        llvm::map_range(llvm::seq<int64_t>(0, rank), [&](int64_t dim) -> Value {
+          AffineExpr expr = rewriter.getAffineDimExpr(0) +
+                            rewriter.getAffineSymbolExpr(0) +
+                            rewriter.getAffineSymbolExpr(1);
+          return linalg::applyMapToValues(
+              rewriter, loc, AffineMap::get(1, 2, expr),
+              {sourceShape[dim], lowPadVals[dim], highPadVals[dim]})[0];
+        }));
+    Value initTensor = rewriter.create<linalg::InitTensorOp>(
+        loc, outputShape, sourceType.getElementType());
+    Value fill = rewriter.create<linalg::FillOp>(loc, initTensor, yieldVal)
+                     ->getResult(0);
+    SmallVector<OpFoldResult> offset = padTensorOp.getMixedLowPad();
+    SmallVector<OpFoldResult, 4> sizes = llvm::to_vector<4>(llvm::map_range(
+        sourceShape, [](Value v) -> OpFoldResult { return v; }));
+    SmallVector<OpFoldResult, 4> strides = llvm::to_vector<4>(llvm::map_range(
+        llvm::seq<int64_t>(0, rank), [&](int64_t dim) -> OpFoldResult {
+          return rewriter.getI64IntegerAttr(1);
+        }));
+    Value subTensorInsert = rewriter.create<SubTensorInsertOp>(
+        loc, source, fill, offset, sizes, strides);
+    rewriter.replaceOpWithNewOp<tensor::CastOp>(
```

Comment:
Why do we need `tensor.cast` here ?

Reply:
`subtensor_insert` deduces the return type. If it doesnt match then you need to cast it. It should match after canonicalization, i.e. if one of the shapes is dynamic at construction time (and the result is all static), you know that it is the same static shape and it needs canonicalizations to fill this up. I think after the shape inference stuff lands this will be a bit more clean. Updated this to not create this if it is not needed in the first place.